### PR TITLE
Enable @Validate as a common behavior

### PR DESCRIPTION
**Pull request by jessemyers**
We want to enable the `@Validate` decorator for any decorator, so we refactor
the shared array decorators into a shared set of decorators of multiple types
that includes both the array and this validate logic.

As part of making this work, I also split the interfaces and options definitions
as the former is mostly internal and the latter is external. I also replaced
`Constructor` with the `@nestjs/common` `Type` declaration (which is equivalent).

### CODE REVIEW

File: src/flavors/__tests__/fixtures.ts [hunk 1]
```diff
@@ -1,6 +1,8 @@
-import { Constructor, Flavor } from '../../interfaces';
```
Comment (jessemyers) [inline src/flavors/__tests__/fixtures.ts:1]: The `Constructor` type was fine, but was also redundant with `Type`. Eliminating our own type definition reduces the surface area of the non-options interfaces, which I think is good.

File: src/recipes/common.recipe.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import { ArrayMinSize, ArrayMaxSize, IsArray, Validate } from 'class-validator';
+
+import { ArraySizeOptions, DecoratorOptions, ValidateOptions } from '../options';
+
+export function buildArrayPropertyDecorators(
```
Comment (jessemyers) [inline src/recipes/common.recipe.ts:5]: This logic was defined in another PR (#32). This file now contains multiple bits of common/shared validation logic.

File: src/recipes/is-array.recipe.ts [hunk 3]
```diff
@@ -1,38 +1,18 @@
...
-
-    // maybe: add a minimum size
-    if (options !== true && options.minItems !== undefined) {
-        decorators.push(ArrayMinSize(options.minItems));
-    }
-
-    return decorators;
-}
-
 export function IsArrayRecipe<Options>(
```
Comment (jessemyers) [inline src/recipes/is-array.recipe.ts:16]: The logic for `isArray` is different from the other decorators; it'd be good to deprecate it now that we have an alternative.

File: src/options/decorator.options.ts [hunk 4]
```diff
@@ -1,3 +1,6 @@
+// eslint-disable-next-line @typescript-eslint/ban-types
+export type ValidateOptions = Function;
```
Comment (jessemyers) [inline src/options/decorator.options.ts:2]: The type definition in `class-validator` is:

```ts
import { ValidationOptions } from '../ValidationOptions';
/**
 * Registers custom validator class.
 */
export declare function ValidatorConstraint(options?: {
    name?: string;
    async?: boolean;
}): (target: Function) => void;
/**
 * Performs validation based on the given custom validation class.
 * Validation class must be decorated with ValidatorConstraint decorator.
 */
export declare function Validate(constraintClass: Function, validationOptions?: ValidationOptions): PropertyDecorator;
export declare function Validate(constraintClass: Function, constraints?: any[], validationOptions?: ValidationOptions): PropertyDecorator;
```

That is, I'm only mapping the `constraintClass` right now.

I'm omitting the secondary `validatorOptions` and `constraints` inputs here for now, but we can generalize `ValidateOptions` to be a union type if we find a need for these extra options.

**Review by adrian-hippo — APPROVED** [review]

**Pull request merged**